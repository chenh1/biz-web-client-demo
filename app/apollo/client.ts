// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});