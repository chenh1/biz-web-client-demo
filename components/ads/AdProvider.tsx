"use client"
import type { AdTags } from "./tags/types"

import { usePathname } from "next/navigation"
import Script from "next/script"
import React, { useEffect, useState, createContext, useRef, Context } from "react"
import { Overlay } from "../containers/Overlay/Overlay"
import { Heading } from "../core/Heading/Heading"
import { Paragraph } from "../core/Paragraph/Paragraph"
import { Box } from "../containers/Box/Box"
import { useFlippers } from "../utils/useFlippers"

const NETWORK_FAILS_THRESHOLD: number = 2
const COLLAPSED_ELEMENT_CHECKS_THRESHOLD: number = 6

export const AdContext: Context<{ adScriptsLoaded: boolean; shouldShow: boolean; }> = createContext({ adScriptsLoaded: false, shouldShow: false } as { adScriptsLoaded: boolean; shouldShow: boolean; })

export interface ProviderContextType {
  adScriptsLoaded: boolean,
  shouldShow: boolean
}

export type AdProviderType = {
  tags: AdTags,
  children?: string | JSX.Element | JSX.Element[], 
}

export const AdProvider: React.FC<AdProviderType> = ({ tags, children }: AdProviderType) => {
  const networkCheckInterval = useRef<NodeJS.Timeout | undefined>(undefined)
  const collapsedElementCheckInterval = useRef<NodeJS.Timeout | undefined>(undefined)
  const pathName = usePathname();
  const [ adScriptsLoaded, setAdScriptsLoaded ]: [boolean, (boolean) => void] = useState(false)
  const pathChanged = useRef(undefined)
  const { ads: shouldShow } = useFlippers({ key: 'ads' });
  

  const adChecks = useRef({ networkFails: 0, collapsedElementChecks: 0 })
  const [ adChecksState, setAdChecksState ] = useState(adChecks.current)

  const updateAdChecksState = (newAdChecks) => {
    adChecks.current = newAdChecks
    setAdChecksState(newAdChecks)
  }
  
  
  useEffect(() => {
    if (typeof window !== 'undefined' && shouldShow) {
      window.googletag = window.googletag || { cmd: [] };

      window.googletag.cmd.push(function() {
        window.googletag.destroySlots?.();

        tags?.forEach(({ elId, path, isResponsive, slots, isIterative }) => {
          if (isResponsive && !adScriptsLoaded) {
            // check if responsive ad slot has already been defined
            if (!isIterative || (isIterative && document.getElementById(elId))) {
              const responsiveAdSlot = window.googletag.defineSlot?.(path, slots, elId).addService(window.googletag.pubads?.());
            
              const mapping = window.googletag.sizeMapping?.()
                .addSize([1024, 768], [[750, 200], [728, 90]])
                .addSize([640, 480], [300, 250])
                .addSize([0, 0], [])
                .build();
              
              responsiveAdSlot?.defineSizeMapping(mapping);
            }
          } else {
            window.googletag
              .defineSlot?.(path, slots, elId)
              .addService(window.googletag.pubads?.());
          }
        })

        window.googletag.enableServices?.();
      })
      
      setAdScriptsLoaded(true)
    }
  }, [ pathName, shouldShow ])

  useEffect(() => {
    if (typeof window !== 'undefined' && shouldShow) {
      networkCheckInterval.current = setInterval(() => {
        fetch(`https://www.googletagmanager.com/gtm.js?id=${process.env.GTM_ID}`)
          .then((res) => {
            clearInterval(networkCheckInterval.current)
          })
          .catch((e) => {
            updateAdChecksState({
              ...adChecks.current,
              networkFails: adChecks.current.networkFails + 1
            })

            if (adChecks.current.networkFails >= NETWORK_FAILS_THRESHOLD) {
              clearInterval(networkCheckInterval.current)
            }
          })
      }, 2000)

      collapsedElementCheckInterval.current = setInterval(() => {
        const adContainer = document.querySelector('.js-ad-container')
        if (!!adContainer && typeof adContainer?.children[0] === 'undefined') {
          updateAdChecksState({
            ...adChecks.current,
            collapsedElementChecks: adChecks.current.collapsedElementChecks + 1
          })
          
          if (adChecks.current.collapsedElementChecks >= COLLAPSED_ELEMENT_CHECKS_THRESHOLD) {
            clearInterval(collapsedElementCheckInterval.current)
          }
        }
      }, 1500)
    }

  }, [])
  
  const shouldShowAdWarningOverlay = adChecks?.current?.networkFails >= NETWORK_FAILS_THRESHOLD && adChecks?.current?.collapsedElementChecks >= COLLAPSED_ELEMENT_CHECKS_THRESHOLD
  
  return (
    <>
      <Script async src="https://securepubads.g.doubleclick.net/tag/js/gpt.js" />
      <AdContext.Provider value={{ adScriptsLoaded, shouldShow }}>
        <Overlay toggleVisibility={() => {}} visibility={shouldShowAdWarningOverlay}>
          <Box gap="sm" className="max-w-[1080px]">
            <Heading type="h3">Hey there, about your ad blocker...</Heading>
            <Paragraph>
              We get it. Ads can be annoying. We're offering you content for free and we have server bills to pay for to keep this up and running. Please take this into consideration. If you're cool with it, please turn it off and refresh the page. If you feel like you absolutely cannot turn off your ad blocker, then please navigate away from this website. Thanks for your time; we appreciate it!
            </Paragraph>
          </Box>
        </Overlay>
        {children}
      </AdContext.Provider>
    </>
  )
}