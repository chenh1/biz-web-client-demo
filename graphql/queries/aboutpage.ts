import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;