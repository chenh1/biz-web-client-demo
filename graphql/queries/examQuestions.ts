import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`