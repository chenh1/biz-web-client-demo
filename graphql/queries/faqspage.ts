import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;