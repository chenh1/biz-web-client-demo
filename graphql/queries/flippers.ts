import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
