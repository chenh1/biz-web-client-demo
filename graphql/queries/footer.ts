import { gql, DocumentNode } from "@apollo/client";

export type FooterData = {
  linkGroups: {
    title: string;
    link: {
      url: string;
      copy: string;
    }[]
  }[];
  disclaimer: string;
}

export type FooterDataRes = {
  footer: {
    data: {
      attributes: FooterData
    }
  }
}

export const footerQuery: DocumentNode = gql`
  query Footer {
    footer {
      data {
        attributes {
          linkGroups {
            link {
              url
              copy
            }
            title
          }
          disclaimer
        }
      }
    }
  }
`