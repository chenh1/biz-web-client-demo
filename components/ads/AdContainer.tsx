"use client"

import React, { useEffect, useContext } from "react";
import { AdContext } from "./AdProvider";
import './adContainer.css';

type AdContainerType = {
  elId: string;
  align: "center" | undefined;
}

export const AdContainer: React.FC<AdContainerType> = ({ elId = "base-ad", align = "center" }) => {
  const { adScriptsLoaded, shouldShow } = useContext<{ adScriptsLoaded: boolean; shouldShow: boolean; }>(AdContext)
  
  useEffect(() => {
    if (typeof window !== 'undefined' && adScriptsLoaded && shouldShow) {
      window.googletag.cmd.push(function() {
        window.googletag.display?.(elId);
      });
    }
  }, [ adScriptsLoaded ])

  return shouldShow
    ? <div className={`ad-wrapper ${align}`}>
        <div className="js-ad-container" id={elId}></div>
      </div>
    : null
}