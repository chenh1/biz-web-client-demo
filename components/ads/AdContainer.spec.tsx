import renderer from 'react-test-renderer';
import { AdContainer } from './AdContainer';
import { AdProvider } from './AdProvider';
import { MockedProvider } from "@apollo/client/testing";
import postTags from '../ads/tags/post'
import { flippersQuery } from '../../graphql/queries/flippers';
import { JSDOM } from "jsdom"
import { render, waitFor } from '@testing-library/react';
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock('next/script', () => {
  // eslint-disable-next-line react/display-name, react/prop-types
  return function ({ children, 'data-testid': dataTestId, ...props }) {
    return (
      <script data-testid={dataTestId} {...props}>
        {children}
      </script>
    );
  };
});

const mocks = [
  {
    request: {
      query: flippersQuery,
    },
    result: {
      data: {
        flipperManager: {
          data: {
            attributes: { ads: true }
          }
        }
      }
    }
  }
];

it('outputs an ad wrapper div', async () => {
  const { getByTestId, asFragment } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AdProvider tags={postTags}>
        <AdContainer align="center" elId="leaderboard1"/>
      </AdProvider>
    </MockedProvider>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});