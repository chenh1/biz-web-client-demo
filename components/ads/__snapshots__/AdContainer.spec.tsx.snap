// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs an ad wrapper div 1`] = `
<DocumentFragment>
  <script
    async=""
    src="https://securepubads.g.doubleclick.net/tag/js/gpt.js"
  />
</DocumentFragment>
`;
